### PR TITLE
feat(papyrus_p2p_sync): impl p2psyncclient internal block forwading

### PR DESCRIPTION
**Pull request by eitanm-starkware**

### CODE REVIEW

**Review by ShahakShama — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2677)* status: all files reviewed, 3 unresolved discussions (waiting on @eitanm-starkware)

___
*[`crates/papyrus_p2p_sync/src/client/mod.rs` line 254 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEhRZsF1vYBDpTWkEHG:-OEhRZsF1vYBDpTWkEHH:b-fa9e9h) ([raw file](https://github.com/starkware-libs/sequencer/blob/5d17ec5b2392ead169f1cc544fce0326fba1b5e3/crates/papyrus_p2p_sync/src/client/mod.rs#L254)):*
> ```XML
>             internal_blocks_receiver,
>         } = self;
>         Self::create_internal_blocks_sender_task(
> ```

You should poll the task handle returned with select
___
*[`crates/papyrus_p2p_sync/src/client/mod.rs` line 270 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEhQRxY8cIXRv63NHjD:-OEhQRxY8cIXRv63NHjE:b-8c5hcl) ([raw file](https://github.com/starkware-libs/sequencer/blob/5d17ec5b2392ead169f1cc544fce0326fba1b5e3/crates/papyrus_p2p_sync/src/client/mod.rs#L270)):*
> ```XML
>     }
> 
>     fn create_internal_blocks_sender_task(
> ```

I think that this should sit in DataStreamBuilder somehow so that each protocol will define to itself how to read data from it

Maybe you can receive a `Receiver<Arc<SyncBlock>>` in create_new_stream and have a generic function that receives an Arc<SyncBlock> and returns Self::Output?

You can add this as a TODO and fix it later

**Review by ShahakShama — CHANGES_REQUESTED** [review]
> Reviewed 6 of 6 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2677)* status: all files reviewed, 9 unresolved discussions (waiting on @eitanm-starkware)

___
*[`crates/papyrus_p2p_sync/src/client/header.rs` line 128 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo71fa789kJKUgEejP:-OEo71fa789kJKUgEejQ:breg8zl) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/header.rs#L128)):*
> ```XML
>         sync_block: SyncBlock,
>     ) -> Option<SignedBlockHeader> {
>         Some(SignedBlockHeader {
> ```

Add TODO to add block info to SyncBlock and use it here
___
*[`crates/papyrus_p2p_sync/src/client/header.rs` line 128 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo7C_QDIVINmMM-A6a:-OEo7C_QDIVINmMM-A6b:b4bgc1f) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/header.rs#L128)):*
> ```XML
>         sync_block: SyncBlock,
>     ) -> Option<SignedBlockHeader> {
>         Some(SignedBlockHeader {
> ```

Add TODO to fill this with real header once SyncBlock contains data required by full nodes
___
*[`crates/papyrus_p2p_sync/src/client/class.rs` line 138 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo6ZnQ6dwjmdlNbmsZ:-OEo6ZnQ6dwjmdlNbms_:b-n5joea) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/class.rs#L138)):*
> ```XML
>         _sync_block: SyncBlock,
>     ) -> Option<(DeclaredClasses, DeprecatedDeclaredClasses, BlockNumber)> {
>         None
> ```

Add TODO to implement this once we have a class manager component and remove the Option in data stream trait
___
*[`crates/papyrus_p2p_sync/src/client/stream_builder.rs` line 61 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo2w4jAwJT9bcdl0wC:-OEo2w4k0A_dng94iCRK:bmhu8hy) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/stream_builder.rs#L61)):*
> ```XML
>         block_number: BlockNumber,
>         sync_block: SyncBlock,
>     ) -> Option<Self::Output>;
> ```

Document what does None mean (It means that we don't want to output anything for this block, and it will usually return when internal blocks are disabled for this stream)
___
*[`crates/papyrus_p2p_sync/src/client/stream_builder.rs` line 81 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo3H1vC6GHNDHm97Sb:-OEo3H1vC6GHNDHm97Sc:bluicn5) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/stream_builder.rs#L81)):*
> ```XML
>                     match Self::convert_sync_block_to_block_data(block_number, sync_block) {
>                         Some(block_data) => block_data,
>                         None => return None,
> ```

This is a bit weird in the case that convert_sync_block_to_block_data returns None for some blocks but not all.

If block_number == current_block_number then return None. Otherwise you should continue the loop

Alternatively, make it clearer in the docs that if you return None for any block it means you don't use internal blocks at all
___
*[`crates/papyrus_p2p_sync/src/client/transaction.rs` line 94 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo6Ono-lThfLTHQk36:-OEo6Ono-lThfLTHQk37:b-w9uhqd) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/transaction.rs#L94)):*
> ```XML
>         sync_block: SyncBlock,
>     ) -> Option<(BlockBody, BlockNumber)> {
>         let num_transactions = sync_block.transaction_hashes.len();
> ```

Add TODO to fill this with real transactions once SyncBlock contains data required by full nodes
___
*[`crates/papyrus_p2p_sync/src/client/transaction.rs` line 99 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo5U1E2jxjf0f4I8TY:-OEo5U1F6wkpeUW6yjjd:b-ruj2ov) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/transaction.rs#L99)):*
> ```XML
>             transaction_hashes: sync_block.transaction_hashes,
>             transaction_outputs: vec>                 TransactionOutput::get_test_instance(&mut rng);
> ```

I'm pretty sure this creates a vector with the same tx num_transactions times. Use std::iter::repeat_with and take
___
*[`crates/papyrus_p2p_sync/src/client/transaction.rs` line 102 at r5 ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/transaction.rs#L102)):*
> ```XML
>                 num_transactions
>             ],
>             transactions: vec![Transaction::get_test_instance(&mut rng); num_transactions],
> ```

Same here

File: crates/papyrus_p2p_sync/src/client/mod.rs [hunk 1]
```diff
@@ -230,14 +240,125 @@ impl P2PSyncClient {
...
+                let num_transactions = sync_block.transaction_hashes.len();
+                let mut rng = get_rng();
+                let block_body = BlockBody {
+                    transaction_hashes: sync_block.transaction_hashes,
+                    transaction_outputs: vec![
+                        TransactionOutput::get_test_instance(&mut rng);
+                        num_transactions
+                    ],
+                    transactions: vec![Transaction::get_test_instance(&mut rng); num_transactions],
+                };
```
[the remark this answers is not in the document]
Comment (eitanm-starkware) [reply crates/papyrus_p2p_sync/src/client/mod.rs]: Done.

**Review by eitanm-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2677)* status: all files reviewed, 9 unresolved discussions (waiting on @graphite-app[bot] and @ShahakShama)

___
*[`crates/papyrus_p2p_sync/src/client/class.rs` line 138 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo6ZnQ6dwjmdlNbmsZ:-OEo8rHc75m7FFYssEJP:b-h6tr1l) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/class.rs#L138)):*
<details><summary><i>Previously, ShahakShama wrote…</i></summary><blockquote>

Add TODO to implement this once we have a class manager component and remove the Option in data stream trait
</blockquote></details>

added half the todo here and half in streambuilder
___
*[`crates/papyrus_p2p_sync/src/client/header.rs` line 128 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo71fa789kJKUgEejP:-OEo9T9RC1SwkluF1wqG:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/header.rs#L128)):*
<details><summary><i>Previously, ShahakShama wrote…</i></summary><blockquote>

Add TODO to add block info to SyncBlock and use it here
</blockquote></details>

Done.
___
*[`crates/papyrus_p2p_sync/src/client/header.rs` line 128 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo7C_QDIVINmMM-A6a:-OEo9cau0GdmY6trpBTJ:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/header.rs#L128)):*
<details><summary><i>Previously, ShahakShama wrote…</i></summary><blockquote>

Add TODO to fill this with real header once SyncBlock contains data required by full nodes
</blockquote></details>

Done.
___
*[`crates/papyrus_p2p_sync/src/client/stream_builder.rs` line 61 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo2w4jAwJT9bcdl0wC:-OEoANiNEunruvPucvDM:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/stream_builder.rs#L61)):*
<details><summary><i>Previously, ShahakShama wrote…</i></summary><blockquote>

Document what does None mean (It means that we don't want to output anything for this block, and it will usually return when internal blocks are disabled for this stream)
</blockquote></details>

Done.
___
*[`crates/papyrus_p2p_sync/src/client/stream_builder.rs` line 81 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo3H1vC6GHNDHm97Sb:-OEoAy-r5kJW-XiRBxjS:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/stream_builder.rs#L81)):*
<details><summary><i>Previously, ShahakShama wrote…</i></summary><blockquote>

This is a bit weird in the case that convert_sync_block_to_block_data returns None for some blocks but not all.

If block_number == current_block_number then return None. Otherwise you should continue the loop

Alternatively, make it clearer in the docs that if you return None for any block it means you don't use internal blocks at all
</blockquote></details>

Done.
___
*[`crates/papyrus_p2p_sync/src/client/transaction.rs` line 94 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo6Ono-lThfLTHQk36:-OEoBJvI76ejGxzZKAbj:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/transaction.rs#L94)):*
<details><summary><i>Previously, ShahakShama wrote…</i></summary><blockquote>

Add TODO to fill this with real transactions once SyncBlock contains data required by full nodes
</blockquote></details>

Done.
___
*[`crates/papyrus_p2p_sync/src/client/transaction.rs` line 99 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo5U1E2jxjf0f4I8TY:-OEoCGH0526S65NNzXjP:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/transaction.rs#L99)):*
<details><summary><i>Previously, ShahakShama wrote…</i></summary><blockquote>

I'm pretty sure this creates a vector with the same tx num_transactions times. Use std::iter::repeat_with and take
</blockquote></details>

Done.
___
*[`crates/papyrus_p2p_sync/src/client/transaction.rs` line 102 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo6LR01hnLEf_sSwoQ:-OEoCGk769XvARLaG1sz:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/transaction.rs#L102)):*
<details><summary><i>Previously, ShahakShama wrote…</i></summary><blockquote>

Same here
</blockquote></details>

Done.

**Review by ShahakShama — CHANGES_REQUESTED** [review]
> Reviewed 7 of 7 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2677)* status: all files reviewed, 5 unresolved discussions (waiting on @eitanm-starkware and @graphite-app[bot])

___
*[`crates/papyrus_p2p_sync/src/client/header.rs` line 128 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo7C_QDIVINmMM-A6a:-OErjaYF9x2EGMijUOLN:bv9yaw2) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/header.rs#L128)):*
<details><summary><i>Previously, eitanm-starkware wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Rephrase it to have the words "contains data required by full nodes"
___
*[`crates/papyrus_p2p_sync/src/client/stream_builder.rs` line 61 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo2w4jAwJT9bcdl0wC:-OErjlAQ8FCrb5WhkEwj:b-8rogs9) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/stream_builder.rs#L61)):*
<details><summary><i>Previously, eitanm-starkware wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Put /// instead of // for documentation
___
*[`crates/papyrus_p2p_sync/src/client/transaction.rs` line 94 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo6Ono-lThfLTHQk36:-OErk6y-7ZhJdMXp-GxE:b36aixk) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/transaction.rs#L94)):*
<details><summary><i>Previously, eitanm-starkware wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Same as before

**Review by eitanm-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2677)* status: 4 of 7 files reviewed, 4 unresolved discussions (waiting on @graphite-app[bot] and @ShahakShama)

___
*[`crates/papyrus_p2p_sync/src/client/header.rs` line 128 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo7C_QDIVINmMM-A6a:-OErr8fs061O9pIvn5DM:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/header.rs#L128)):*
<details><summary><i>Previously, ShahakShama wrote…</i></summary><blockquote>

Rephrase it to have the words "contains data required by full nodes"
</blockquote></details>

Done.
___
*[`crates/papyrus_p2p_sync/src/client/stream_builder.rs` line 61 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo2w4jAwJT9bcdl0wC:-OErr9QS6j9MO_9ohcib:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/stream_builder.rs#L61)):*
<details><summary><i>Previously, ShahakShama wrote…</i></summary><blockquote>

Put /// instead of // for documentation
</blockquote></details>

Done.
___
*[`crates/papyrus_p2p_sync/src/client/transaction.rs` line 94 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-OEo6Ono-lThfLTHQk36:-OErr9ku29KSDoPzu0mY:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/85a6adcaa82e5cd59865c0428265c8e57dcfc5aa/crates/papyrus_p2p_sync/src/client/transaction.rs#L94)):*
<details><summary><i>Previously, ShahakShama wrote…</i></summary><blockquote>

Same as before
</blockquote></details>

Done.

**Review by ShahakShama — APPROVED** [review]
:lgtm:

> Reviewed 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-:-OEru1heAGU36i0rhoyB:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @graphite-app[bot])

**Review by eitanm-starkware — COMMENTED** [review]
> Dismissed @graphite-app[bot] from a discussion.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2677)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @eitanm-starkware)

**Review by ShahakShama — APPROVED** [review]
:lgtm:

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2677#-:-OErv-ZE1buno2g9WPsZ:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @eitanm-starkware)
